### PR TITLE
Show Indels and Soft Clip alignments

### LINKED ISSUE #85: Show indels in BAM alignments
Cigar Ops to support:

- [ ] `I` (insert)
- [ ] `D` (delete)
- [ ] `S` (soft clip)

### LINKED ISSUE #111: Handle alignments without CIGAR strings
The highlighted read lacks a CIGAR string (it's set to `*` in the SAM file):

screen shot 2015-04-28 at 11 12 24 am

IGV shows its mate, but not the paired read.

```
$ samtools view test/data/synth3.normal.17.7500000-7515000.bam | grep 8adfa993-c734-4471-a56c-0c42246967a3
8adfa993-c734-4471-a56c-0c42246967a3	73	17	7512447	37	101M	=	7512447	0	ACAAAAATAAAAGAATTAGAGGTTTCAAAAGGCTAATACACAAGAACTATTCATGAGCTTTGCTGCTAGGGACAAAGGCTAAGCTTAGGAGAGAAAACACC	@@AACDD@ADDDDCDBCADCDDAB@CBEEEECDDBDB@BBA>ADCEBDBBCDBBDCEDDCCDDDDDDBDDDDBBCDDDECBCEB<??C>ACCCCCDBBABB	MD:Z:101	PG:Z:bwa.1	RG:Z:cb0d682c-1a04-47b6-b380-02a3258d7906	AM:i:0	NM:i:0	SM:i:37	OQ:Z:@?@DDFFDHHFHHIIIGEGCHHFHDDGIIIIEGGGGGDGGB<BFGHGGHIIIIIIGFGHGGGEHGIGEGGGE@HAEEFD@C@C;.;>A>;?B@BAC?CBAB	UQ:i:0
8adfa993-c734-4471-a56c-0c42246967a3	133	17	7512447	0	*	=	7512447	0	AGGACGGGGGGGGGCAGAGGAAATTTAAAGATTGGGGCGGAATTAAGGACTAAAAAAAGACAATTAGTCAAGCAAGAGGGTGATCGACTGCTGCCTCTGTG	#####################################################################################################	RG:Z:cb0d682c-1a04-47b6-b380-02a3258d7906	OQ:Z:#####################################################################################################
```

I'm not sure what the significance of this is. Is it unmapped? It has a sequence.

### PR DESCRIPTION
**Pull request by danvk**
This also handles mismatches in the presence of more complex Cigar strings.

Fixes #85 
Fixes #111

pileup indels

[Review on Reviewable](https://reviewable.io/reviews/hammerlab/pileup.js/157)

### CODE REVIEW

File: src/PileupTrack.js [hunk 1]
```diff
@@ -62,6 +61,65 @@ function makePath(scale, visualRead: VisualAlignment) {
   return d3.svg.line()(path);
 }
 
+// Create the SVG element for a single Cigar op in an alignment.
+function enterSegment(parentNode, op, scale) {
+  if (op.op == 'M') {
```
Comment (ihodes) [inline src/PileupTrack.js]: Could be good to comment what 'M', 'D', etc are. I'm guessing match, deletion, insertion? Could also use const variables for them/remove magic strings.

File: src/PileupTrack.js [hunk 2]
```diff
@@ -222,7 +273,8 @@ class NonEmptyPileupTrack extends React.Component {
 
     var referenceSource = this.props.referenceSource;
     var vReads = this.state.reads.map(
-        read => this.addRead(read, referenceSource));
+            read => this.addRead(read, referenceSource))
+        .filter(read => read.refLength);  // drop alignments w/o CIGARs
```
Comment (ihodes) [inline src/PileupTrack.js]: Would it be useful to warn or report on this?

File: src/pileuputils.js [hunk 3]
```diff
@@ -95,8 +86,95 @@ function getDifferingBasePairs(read: SamRead, reference: string): Array<BasePair
...
+      mismatches = mismatches.concat(findMismatches(ref, mSeq, refPos));
+    }
+
+    result.push({
+      op: op.op,
+      length: op.length,
+      pos: refPos
+    });
+
+    switch (op.op) {
```
Comment (ihodes) [inline src/pileuputils.js]: Same magic strings comment here and below.

Comment (ihodes) [inline src/pileuputils.js]: (I realize it's CIGAR parsing, but it'd help to have it briefly documented here, anyway—your call though)

**Comment by ihodes** [comment]
While not the most d3-y, (though maybe it is the *most* d3-y you can get in this situation?) this doesn't seem bad to me at all. 

In terms of making it a little more readable I'd *maybe* have gone for a case/switch on the op, and factor out the body of each to a function, but this is pretty manageable to read and understand as it is, with the 3 (+1) cases. 

Comments on what each branch is doing (e.g. just `// D: deletion, make a thin rectangle`) might be helpful coming back to this, but just having the 'D' be in a var like `DELETION_CIGAR_OP` would be more than sufficient, too.

Anyway, this all looks really fine, comments are all nits/minor suggestions.

File: src/PileupTrack.js (hunk 1 above)
Comment (danvk) [inline src/PileupTrack.js]: I factored out a `CigarOps` enum. It's much clearer now.

File: src/PileupTrack.js (hunk 2 above)
Comment (danvk) [inline src/PileupTrack.js]: It's actually fairly common (see #111). I'm not sure what alignments without Cigar strings correspond to—maybe @arahuja or @smondet know more.

**Comment by danvk** [comment]
I factored out a `CigarOps` enum and changed to `switch`/`case`. I also moved the drawing code to all be in `updateSegment` and added comments indicating what's being drawn.

**Pull request merged**